pmlc/dialect/comp/transforms/minimize_buffer_transfers.cc

```diff
@@ -0,0 +1,154 @@
+      for (auto &use : opResult.getUses()) {
+        uses.insert(&use);
+      }
+    }
+  }
+
+  for (auto operation : operations) {
+    if (isInternalOperation(operation)) {
+      operations.erase(operation);
+    }
+  }
+  return operations;
+}
+
+bool MinimizeBufferTransfersPass::isInternalOperation(mlir::Operation *op) {
+  auto allocOp = llvm::dyn_cast<mlir::AllocOp>(op);
+  auto memRefCastOp = llvm::dyn_cast<mlir::MemRefCastOp>(op);
+  auto waitOp = llvm::dyn_cast<comp::Wait>(op);
+  if (allocOp || memRefCastOp || waitOp) {
+    return true;
+  }
+  return false;
+}
+
+BufferCopyMode
+MinimizeBufferTransfersPass::getBufferCopyMode(mlir::Operation *op,
+                                               mlir::Value buffer) {
+  BufferCopyMode copyMode = BufferCopyMode::NoCopy;
+  if (buffer.isa<mlir::BlockArgument>()) {
+    copyMode = copyMode |
+               (BufferCopyMode::HostToDevice | BufferCopyMode::DeviceToHost);
+    return copyMode;
+  }
+
+  auto currentBlock = op->getBlock();
+  auto operationDeps = getExternalDependentOperations(buffer);
+  for (auto dependency : operationDeps) {
+    if (dependency->getBlock() == currentBlock) {
+      if (dependency->isBeforeInBlock(op)) {
+        copyMode = copyMode | BufferCopyMode::HostToDevice;
+      }
+      if (op->isBeforeInBlock(dependency)) {
+        copyMode = copyMode | BufferCopyMode::DeviceToHost;
+      }
+    } else {
+      // Dependency is outside of current block.
+      copyMode = copyMode |
+                 (BufferCopyMode::HostToDevice | BufferCopyMode::DeviceToHost);
+    }
+  }
+  return copyMode;
+}
+
+void MinimizeBufferTransfersPass::removeWaitOpEvent(comp::Wait waitOp,
+                                                    mlir::Value event) {
+  auto events = waitOp.events();
+  for (size_t i = 0; i < events.size(); i++) {
+    auto pEvent = std::next(events.begin(), i);
+    if (*pEvent == event) {
+      waitOp.eventsMutable().erase(i);
+      break;
+    }
+  }
+}
+
+void MinimizeBufferTransfersPass::runOnFunction() {
+  auto func = getFunction();
+
+  func.walk([&](comp::ScheduleWrite op) {
+    auto copyMode = getBufferCopyMode(op.getOperation(), op.hostMem());
+    if ((copyMode & BufferCopyMode::HostToDevice) == BufferCopyMode::NoCopy) {
+      auto users = op.getResult().getUsers();
+      for (auto user : users) {
+        if (mlir::isa<comp::Wait>(user)) {
+          removeWaitOpEvent(mlir::cast<comp::Wait>(*user), op.getResult());
+        }
+      }
+      op.replaceAllUsesWith(op.deviceMem());
+      op.erase();
+    }
+  });
+
+  func.walk([&](comp::ScheduleRead op) {
+    auto copyMode = getBufferCopyMode(op.getOperation(), op.hostMem());
+    if ((copyMode & BufferCopyMode::DeviceToHost) == BufferCopyMode::NoCopy) {
+      auto users = op.getResult().getUsers();
+      for (auto user : users) {
+        if (mlir::isa<comp::Wait>(user)) {
+          removeWaitOpEvent(mlir::cast<comp::Wait>(*user), op.getResult());
```

Comment:
These sorts of things should do casting once, like:

```
if (auto waitOp = mlir::dyn_cast<comp::Wait>(user)) {
  removeWaitOpEvent(waitOp, op.getResult());
}
```

```

Reply:
I notice this pattern can be applied in much of this PR.

Reply:
Fixed